File: calculate_fasta_median_length.pl
# !/usr/local/bin/perl -w
# calculate_fasta_median_length.pl
# Geoffrey Hannigan
# Elizabeth Grice Lab
# University of Pennsylvania

# Set use
use strict;
use warnings;

# Set files to scalar variables
my $usage = "Usage: perl $0 <INFILE>";
my $infile = shift or die $usage;
open(my $IN, "<", "$infile") || die "Unable to open $infile: $!";

# Set variable for list of length values
my @lengths = ();
my $seq_length = -1;
my $result = 0;

# Assign lengths of each sequence to the array
while (my $line = <$IN>) {
	# Skip the sequence identifier lines
	if ($line =~ /\>/) {
		if ($seq_length != -1) {
			push @lengths, $seq_length;
		}
		$seq_length = 0;
		next;
	} else {
		# Add  line lengths to array
		chomp $line;
		$seq_length += length $line;
	}
}
if ($seq_length != -1) { # do not forget the last record
	push @lengths, $seq_length;
}

# Sort the length array
my @sorted_lengths = sort {$a <=> $b} @lengths;
# Assign array length to a variable
my $array_length = @sorted_lengths;

# Get the median of the array, depending on whether the link is even or odd
if ($array_length%2) {
	# Odd
	my $result = $sorted_lengths[($array_length/2)];
	printf $result."\t".$infile."\n";
} else {
	# Even
	my $one = $sorted_lengths[($array_length/2)-1];
	my $two = $sorted_lengths[($array_length/2)];
	my $result = ($one + $two) / 2;
	printf $result."\t".$infile."\n";
}

#Close out files
close($IN);

File: calculate_fasta_median_length.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "usage: %s <FASTAFILE>\n\n", os.Args[0])
		os.Exit(1)
	}

	file := os.Args[1]
	fh, err := os.Open(file)
	if err != nil {
		fmt.Fprintf(os.Stderr, "fail to open file: %s\n", file)
		os.Exit(1)
	}
	defer func() {
		fh.Close()
	}()

	var line string
	length := -1
	lengths := make([]int, 0, 10000) // slice with initial capacity of 10000

	reader := bufio.NewReader(fh)
	for {
		line, err = reader.ReadString('\n')
		if err == io.EOF {
			break
		}
		if line[0] == '>' { // ignore header line
			if length != -1 { // not the first sequence
				lengths = append(lengths, length)
			}
			length = 0
			continue
		}

		if line[len(line)-1] == '\n' { // remove '\n'
			line = line[0 : len(line)-1]
		}
		if line[len(line)-1] == '\r' { // remove '\r'
			line = line[0 : len(line)-1]
		}
		length += len(line)
	}
	if length != -1 { // do not forget the last record
		lengths = append(lengths, length)
	}

	sort.Ints(lengths) // sort
	// fmt.Println(lengths)

	n := len(lengths)
	if n == 0 {
		fmt.Printf("%d\t%s\n", 0, file)
	} else if n%2 == 1 { // odd
		fmt.Printf("%d\t%s\n", lengths[(n-1)/2], file)
	} else { // even
		fmt.Printf("%d\t%s\n", (lengths[(n-1)/2]+lengths[n/2])/2, file)
	}
}
